### ISSUE TITLE
Error: protect(): protection stack overflow

### ISSUE THREAD

**Issue by xiaguoxin** [rstudio/rstudio#5546]
### System details
    RStudio Edition : Desktop
    RStudio Version : Version 1.2.1335
    OS Version      : macOS 10.14.5
    R Version       : 3.6.1
    &
    RStudio Edition : Server
    RStudio Version : (might be) 1.1.463
    OS Version      : Ubuntu Server 18.04.3 LTS
    R Version       : 3.6.0
### Steps to reproduce the problem
From [ fast single item lookup from list vs data.table vs hash
 ](https://stackoverflow.com/questions/34232865/r-fast-single-item-lookup-from-list-vs-data-table-vs-hash)
```
# Set seed to 42 to ensure repeatability
set.seed(42)

# Generate product ids
product_ids <- as.vector(
  outer(LETTERS[seq(1, 26, 1)],
        outer(outer(LETTERS[seq(1, 26, 1)], LETTERS[seq(1, 26, 1)], paste, sep=""),
              LETTERS[seq(1, 26, 1)], paste, sep = ""
        ), paste, sep = ""
  )
)

# Create test lookup data
test_lookup_list <- lapply(product_ids, function(id){
  return_list <- list(
    product_id = id,
    val_1 = rnorm(1),
    val_2 = rnorm(1),
    val_3 = rnorm(1),
    val_4 = rnorm(1),
    val_5 = rnorm(1),
    val_6 = rnorm(1),
    val_7 = rnorm(1),
    val_8 = rnorm(1)
  )
  return(return_list)
})

# Set names of items in list
names(test_lookup_list) <- sapply(test_lookup_list, function(elem) elem[['product_id']])

test_lookup_env <- list2env(test_lookup_list)
```
### Describe the problem in detail
the sessionInfo:
```R version 3.6.0 (2019-04-26)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Debian GNU/Linux 9 (stretch)

Matrix products: default
BLAS/LAPACK: /usr/lib/libopenblasp-r0.2.19.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8   
 [6] LC_MESSAGES=C              LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] compiler_3.6.0 tools_3.6.0    packrat_0.5.0 
```
I've searched for that and there were many suggestion about start R using argument `--max-ppsize=5000000`, but it seems not working with many post complained, and I don't know how to do it for a RStudio Server.

If there's some suggestion, it will be very helpful.
Thank you.

**Comment by dmurdoch** [comment]
This bug has shown up in https://stackoverflow.com/q/59414426/2554330 with a very simple way to reproduce:

    h <- hash::hash(1:120000, 1:120000)
    # Error: protect(): protection stack overflow

**Comment by pnacht** [comment]
Unsure whether to post this as a separate issue, but using the oneliner above and running it twice causes RStudio to soft-crash.

After the first call, we get the `protect()` error once.

After repeating the call (immediately or after waiting an arbitrary amount of time), we get an infinite sequence of 

```
h <- hash::hash(1:120000, 1:120000)
# Error: protect(): protection stack overflow
h <- hash::hash(1:120000, 1:120000)
# Error: protect(): protection stack overflow
# Error: promise already under evaluation: recursive default argument reference or earlier problems?
# Error: protect(): protection stack overflow
# Error: promise already under evaluation: recursive default argument reference or earlier problems?
# Error: protect(): protection stack overflow
# Error: promise already under evaluation: recursive default argument reference or earlier problems?
```

This effectively crashes RStudio: subsequent calls are ignored, "Restart R" makes the "Restarting R..."  tab thing appear at the top of the screen and stay there, with no progress. and we can't close RStudio normally, only via the Task Manager.

**Comment by kevinushey** [comment]
Note: `bigmemory` is actually one case where this null pointer examination may be warranted: https://github.com/rstudio/rstudio/issues/8923

But invalid `bigmemory` objects are really quite brittle (even asking them how long they are will cause a segfault).

**Comment by astayleraz** [comment]
Verified in Version 1.5.192-1 ubuntu20 that I am not seeing this show `Error: protect(): protection stack overflow` if I have null external pointers set to off in the command palette. Closing issue.

**Comment by 3fmeister** [comment]
Hi guys,

I recently ran into a problem similar to the one described in this issue. Let's say that for reasons of reproducibility we use the aforementioned hash function and the following call:

`h <- hash::hash(1:120000, 1:120000)`

As it was said in the comments above, this leads to an infinite loop of errors appearing in the console, and therefore R Studio freezes.

The error that is printed: **Error: protect(): protection stack overflow**

If I reduce the range for hash function arguments (key-value pairs):

**h <- hash::hash(1:20000, 1:20000)**

There are no more errors, and the variable (environment) **h** is visible in the environment pane.

I understand that the problem is probably the size of the pointer protection stack, but I don't know how to solve it on Mac.

I also came across a suggestion that using the `—max-ppsize` argument when starting R Studio can help in such situations. I'm not sure how or if I can use it on a Mac though. Is there a way to run R Studio on Mac with an increased protection stack, say set to 500000? Would that help at all?

Or maybe the problem is in inefficient and unoptimized code? I'm not just referring to this example I mentioned (it's there only for reproducibility), but in general when a protection stack overflow error occurs.

**System details:**
RStudio Edition: Desktop
RStudio Version: 2022.07.2+576
OS Version: macOS Big Sure (11.6)
R Version: 4.2.0
RAM Memory: 16 GB
